From 391184794d71e3b87f3db5b1b38915dd48bd7917 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Tue, 6 Aug 2024 18:05:45 +0200
Subject: [PATCH 671/841] fix(stm32-rng): RNG doesn't take into account fully
 disabled node
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When no RNG node has a status set to okay the function return an error.
This behavior is not good as RNG could be unused in some BL2
configuration. To avoid this issue we print a warning instead of
returning an error.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git fb7f3e5eb48c97b42d675eab52ce88eea80c01b8 fix(stm32-rng): RNG doesn't take into account fully disabled node]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: I3fbf9480d9c75d52393865013bb0bb4b4253e8fd
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/397151
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/crypto/stm32_rng.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index 57b9ee9b1..4f005a69d 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -235,6 +235,7 @@ int stm32_rng_init(void)
 	void *fdt;
 	int node;
 	int success = 0;
+	int disabled = 0;
 
 	if (stm32_rng.base != 0U) {
 		if(stm32_rng.clock != 0U) {
@@ -263,6 +264,7 @@ int stm32_rng_init(void)
 		 * instance of the driver.
 		 */
 		if ((dt_rng.status == DT_DISABLED) || (dt_rng.base == 0U)) {
+			disabled++;
 			continue;
 		}
 
@@ -303,12 +305,15 @@ int stm32_rng_init(void)
 		}
 	}
 
-	if (success > 0) {
+	if ((success == 0) && (disabled > 0)) {
+		WARN("%s: No RNG found in device tree.\n", __func__);
+		return 0;
+	} else if (success > 0) {
 		if ((rng.clock != 0U) && (rng.base != 0U)) {
 			stm32_rng = rng;
 		}
 		return 0;
+	} else {
+		return -ENODEV;
 	}
-
-	return -ENODEV;
 }
-- 
2.47.2

